### PR TITLE
Directly transmit predictions in parallel instead of implicitly pickling them

### PR DESCRIPTION
**Pull request by iMurfyD**
`mpi4py` lowercase methods implicitly pickle all data before using them in MPI. This uses the uppercase `mpi4py` methods, which directly uses MPI with no modifications to data, to transmit the predictions in order to overcome a memory problem when creating multiple predictions of size >1. MPI can only transmit 1D arrays, so the multidimensional prediction `numpy.array`'s need to be flat packed before shipping. 

If it fits it ships. One low rate, anywhere in the nation.

### CODE REVIEW

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by iMurfyD** [comment]
Tested latest commit with 4 cores for 4 hours on Traverse. Nothing obviously breaking.

**Comment by iMurfyD** [comment]
Previous run of FRNN before this path yielded the following ROC curve 

```
epoch 60, test_roc_30 = 0.6952110289965381 test_roc_70 = 0.6622922242137053 test_roc_200 = 0.6426808907657934 test_roc_500 = 0.6200066291831373 test_roc_1000 = 0.5812320459623364
=========Summary======== for epoch 59.64
Training Loss numpy: 3.015e-01
Validation Loss: 4.357e-01
Validation ROC: 0.8893
========================
```

Using the same `conf.yaml` and training for about the same amount of time, with the patch, FRNN yields this ROC information:

```
epoch 74, test_roc_30 = 0.688759851702718 test_roc_70 = 0.6834196763976528 test_roc_200 = 0.6697439171106583 test_roc_500 = 0.6425826806452404 test_roc_1000 = 0.6132853740577966
=========Summary======== for epoch 73.84
Training Loss numpy: 3.113e-01
Validation Loss: 4.549e-01
Validation ROC: 0.8949
========================
No improvement, saving model weights anyways
Finished evaluation of epoch 73.84/1000
Begin training from epoch 73.84/1000
Compilation finished in 0.57s
```

Further up at a closer epoch:

```
epoch 55, test_roc_30 = 0.7432142944830465 test_roc_70 = 0.6997747305359818 test_roc_200 = 0.6629397971961011 test_roc_500 = 0.6289897861474625 test_roc_1000 = 0.5856177416582778
=========Summary======== for epoch 54.91
Training Loss numpy: 3.155e-01
Validation Loss: 4.694e-01
Validation ROC: 0.9001
========================
Finished evaluation of epoch 54.91/1000
Begin training from epoch 54.91/1000
Compilation finished in 0.57s
```

These ROC characteristics match pretty closely, which probably means that this patch didn't affect training.

File: plasma/models/mpi_runner.py [hunk 1]
```diff
@@ -822,6 +912,11 @@ def mpi_make_predictions_and_evaluate(conf, shot_list, loader,
     y_prime, y_gold, disruptive = mpi_make_predictions(
         conf, shot_list, loader, custom_path)
     analyzer = PerformanceAnalyzer(conf=conf)
+    #g.write_all('Afterwards y_prime length = {}\n'.format(len(y_prime)))
```
Comment (felker) [inline plasma/models/mpi_runner.py]: Can you remove these debug comments?

File: plasma/models/mpi_runner.py [hunk 2]
```diff
@@ -792,19 +827,74 @@ def mpi_make_predictions(conf, shot_list, loader, custom_path=None):
...
+            y_goldg_flattend  = np.zeros(y_goldg.flatten().shape)
+            if color == 1:
+                # Ensure that numpy arrays have correct dimensions before gathering them
+                assert num_pred*max(y_prime_numpy.flatten().shape) == max(y_primeg_flattend.shape)
+                assert num_pred*max(y_gold_numpy.flatten().shape) == max(y_goldg_flattend.shape)
+                temp_predictor_only_comm.Barrier()
+                temp_predictor_only_comm.Allgather(y_prime_numpy.flatten(), y_primeg_flattend)
+                temp_predictor_only_comm.Allgather(y_gold_numpy.flatten(), y_goldg_flattend)
+                temp_predictor_only_comm.Barrier()
+            # Process 0 broadcast y_primeg adn y_goldg to all processors, including ones
```
Comment (felker) [inline plasma/models/mpi_runner.py]: "adn" typo

File: plasma/models/mpi_runner.py [hunk 3]
```diff
@@ -792,19 +827,74 @@ def mpi_make_predictions(conf, shot_list, loader, custom_path=None):
...
+                g.write_all(' max length = {}\n'.format(max_length))
+            g.comm.Barrier()
+            max_length = g.comm.allreduce(max_length, MPI.MAX) 
+            if color == 1:
+                y_prime_numpy = np.stack([np.pad(sublist, pad_width=((0,max_length-max(sublist.shape)),(0,0))) for sublist in y_prime])
+                y_gold_numpy = np.stack([np.pad(sublist, pad_width=((0,max_length-max(sublist.shape)),(0,0))) for sublist in y_gold])
+            
+            g.comm.Barrier()
+            temp_predictor_only_comm = MPI.Comm.Split(g.comm, color, i)
             g.comm.Barrier()
```
Comment (felker) [inline plasma/models/mpi_runner.py]: are these MPI Barriers necessary? I would remove all that arent absolutely critical for correctness

File: plasma/models/mpi_runner.py [hunk 4]
```diff
@@ -792,19 +827,74 @@ def mpi_make_predictions(conf, shot_list, loader, custom_path=None):
...
+            # Process 0 broadcast y_primeg adn y_goldg to all processors, including ones
+            # not involved in calculating predictions so they can each create their own 
+            # y_prime_global and y_gold_global
+            g.comm.Barrier()
+            g.comm.Bcast(y_primeg_flattend, root=0)
+            g.comm.Bcast(y_goldg_flattend, root=0) 
+            g.comm.Barrier()
+            y_primeg_flattend = np.split(y_primeg_flattend, num_pred)
+            y_goldg_flattend = np.split(y_goldg_flattend, num_pred)
+            y_primeg = [y.reshape((128, max_length, 1)) for y in y_primeg_flattend]
```
Comment (felker) [inline plasma/models/mpi_runner.py]: What is the hardcoded 128 here? Should it be a tunable parameter from `conf.yaml`?

File: plasma/models/mpi_runner.py (hunk 2 above)
Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: Fixed

File: plasma/models/mpi_runner.py (hunk 1 above)
Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: Removed these - thank you for catching!

File: plasma/models/mpi_runner.py (hunk 4 above)
Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: This should be the batch size from the `conf.yaml`. I'll change it to that and run a test run on Traverse

Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: Changed and seems to work

File: plasma/models/mpi_runner.py (hunk 3 above)
Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: Probably not. I'll take out the ones which probably don't need to be there and test it

Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: Took unessential Barrier's out. Tested with 4 nodes on Traverse. Nothing seems to be crashing

Comment (iMurfyD) [reply plasma/models/mpi_runner.py]: @felker, if you're satisfied can you resolve the conversation? If not, what else should happen to test this?

**Comment by felker** [comment]
💯 🔥 🚀 **LGTM**

**Pull request merged**